### DISCUSSION TITLE
Get albums for a specific artist

### DISCUSSION THREAD

**Discussion by xiaopeiwu** [johnwmillr/LyricsGenius#189] [Q&A] [answered]
Hi! Thank you for this great package. I've tried a few simple searches and am looking in the documentation, but I can't seem to find a method to get the albums for a specific artist based on the list of songs returned. I'm seeing that although the Artist object has a song attribute, the Song object doesn't have an album attribute. However when I dump the Song object into json I do see album info in there.

Is there a reason for this? Or am I simply missing something obvious? I'm trying to build a dataset for lyrics of an artist where album may be a good feature to have. 

Many thanks!

**Comment by allerter** [comment, marked as answer]
Hi. To get an Artist's albums you can use the `Genius.artist_albums` method and then get more information from there:
```python
albums = genius.artist_albums(45)['albums']
for album in albums:
    full_album = genius.search_album(album_id=album['id'], get_full_info=True)
```
As for why some attributes are present in the dict, but not in the object, it's because of consistency. Depending on the value of the `get_full_info` parameter there will be more or fewer data in the response, but LyricsGenius has to make sure that the values are consistent, so some of them have been removed from the object (but still available through `to_dict` if you want them). We'll probably change this in the future to provide all possible values.

**Reply by xiaopeiwu** [reply]
Hi, thanks for getting back to me! While the `artist_albums` method works well, I guess my original intent was more to obtain a list of albums for that artist based on the song returned using the `search_artist` method - since using that method you can specify how many songs to return and based on popularity or title. If a song is not returned, I'm not really interested in the album. Hope that makes sense!
And thanks for pointing me to `to_dict` as I was able to use that for a workaround to get the album info quickly.
It makes sense that consistency is important, and thanks for explaining the reasoning for that. It's interesting that Album isn't consistently returned in the response for a Song object - I would think that's pretty essential information.
I'd be happy to help with any future improvement of the project!

**Reply by allerter** [reply]
I'm not sure I understand what you mean. How would we get a list of albums based on a song? All `search_artist` does is use `genius.artist_songs` (which has a `sort` parameter) to get artist's songs and in turn, fetch lyrics and info for each song.

If you'd like to contribute to this particular problem, I'd suggest starting by reading [this comment](https://github.com/johnwmillr/LyricsGenius/pull/181#issuecomment-776850319). I laid out the two ways that it can be resolved. Personally, I would remove `get_full_info` altogether. If you'd like to work on this solution, remove the `get_full_info` parameter and add the missing attributes (like `Song.album`) to the `Song` object. If you have any questions regarding what I just explained, open an issue and we'll talk more about it.

**Reply by xiaopeiwu** [reply]
Sorry my bad - I meant _songs_ returned, not a single song. The list (or set in my current code) would contain only the albums of the songs returned by `search_artist` or `artist_songs` - so e.g. if I specify max 3 songs by Radiohead sorted by popularity, and they happen to be in 2 different albums then I get a list of these 2 albums. I'm not interested in any other Radiohead albums.
Sure I'll check out your comment in #181 and get back to you on that.

**Reply by allerter** [reply]
In that case, you could try something like this out:
```python
artist = genius.search_artist("Radiohead", max_songs=3, get_full_info=True)
album_ids = set(song.to_dict()['album']['id'] for song in artist.songs)
```

**Reply by xiaopeiwu** [reply]
Thanks! That's basically what I have already. Just another question regarding `get_full_info`: is it safe to assume that when setting this to `True`, the returned song from the API will always have the album info? I currently check for it as below:
```
album_ids = set()
artist = genius.search_artist("Radiohead", max_songs=3, get_full_info=True)
songs = artist.songs
for song in songs:
    album = song.to_dict().get("album")
        if album:
            album_id = album.get("id")
            if album_id:
                album_ids.add(album_id)
```

**Reply by allerter** [reply]
When `get_full_info` is set, the package uses `genius.song` which is one of the developers' API methods. The return value of these methods is consistent and won't change without prior notice. And as a rule of thumb, if it's there once, it'll be there next time; just how you can be sure that `album['id']` will always be present when there's an album.

**Reply by xiaopeiwu** [reply]
Thanks for all your help! I did a bit more testing and it appears that there is a difference between using the `genius.song` method directly and using `search_song` or `search_artist` while setting `get_full_info=True` and then use `to_dict()`. However, this thread more than answered my original question so I will mark it as answered and maybe open an issue separately. Many thanks again!

**Reply by allerter** [reply]
As far as I can remember the only difference between `genius.song` and `genius.search_song(get_full_info=True).to_dict` is that the latter adds two values: `artist` which is the artist name and `lyrics` which contains the lyrics.

**Reply by xiaopeiwu** [reply]
I've made a small test case to show the difference:
```python
genius = lyricsgenius.Genius(timeout=20, sleep_time=5)  # getting access token from an environment variable
test_song = genius.song(1081664)
test_full_info = genius.search_song("Wonderwall", "Radiohead", get_full_info=True).to_dict()
print(len(test_song["song"]))
print(len(test_full_info))
```
These are the same song. `test_full_info` is missing quite a few values that are present in `test_song`. So I've used the debugger and it appears that in line 430 of `genius.py`, there is this condition:
```python
        if song_id is None and get_full_info is True:
            new_info = self.song(song_id)['song']
            song_info.update(new_info)
```
And since the song does have a `song_id`, this condition evaluates to `False`, and so the update using `genius.song` to the public API call result never happened.

Package version is 3.0.0

**Reply by allerter** [reply]
That's been resolved in #181.

**Reply by xiaopeiwu** [reply]
Oh I see! I should've checked out #181 like I said I would :)
